src/arch/utils.cpp

```diff
@@ -0,0 +1,27 @@
+#include "arch.h"
+
+#include <cstring>
+#include <sys/utsname.h>
+
+namespace bpftrace {
+namespace arch {
+
+int get_kernel_ptr_width()
+{
+  // We can't assume that sizeof(void*) in bpftrace is the same as the kernel
+  // pointer size (bpftrace can be compiled as a 32-bit binary and run on a
+  // 64-bit kernel), so we guess based on the machine field of struct utsname.
+  // Note that the uname() syscall can return different values for compat mode
+  // processes (e.g. "armv8l" instead of "aarch64"; see COMPAT_UTS_MACHINE), so
+  // make sure this is taken into account.
+  struct utsname utsname;
+  if (uname(&utsname) >= 0)
+  {
+    if (!strncmp(utsname.machine, "armv", 4) && utsname.machine[4] < '8')
+      return 32;
+  }
+  return 64;
+}
```

Comment:
Should this be defined in the per-arch implementations instead? This function seems to be making a global assumption that the only 32-bit arch support we have is on arm (which is technically true for now).

Seems like it'd be cleaner to do it per-arch and have the 64-bit archs just return 64. So that `get_kernel_ptr_width()` is implemented like the rest of the helpers declared in `arch.h`.

Reply:
Good idea, will update in a bit!